System workqueue: Prevent blocking API calls

bjarki-andreasen opened the pull request:
It is inherently unsafe to call blocking APIs from the system work queue, as it is shared opaquely between modules which may use the system work queue internally to unblock the blocking API call (deadlock).

This particular deadlock is extraordinarily hard to hunt down, given there is no hard fault, and it affects seemingly random parts of the system. Drivers and subsystems just "stop responding" from a seemingly arbitrary, unrelated API call.

Subsystems which rely heavily on the system workqueue already adhere to the fact that blocking API calls are unsafe, and have implemented their own workarounds to address this.

This PR addresses the issue by extending the documentation to state that it is inherently unsafe to use the system work queue for blocking API calls, and introduces an optional check similar to a stack sentinal or spinlock verify, which invokes a kernel oops if a work item passed to the system work queue attempts a blocking call. Lastly, the new kernel API `k_is_in_sys_work()` for checking if a calling context is within a work item handler passed to the system workqueue is added.

**Note** the new API is `k_is_in_sys_work()`, not `k_is_in_sys_workqueue_thread()`. The thread is allowed to unready when it is not busy, between items for example, it is only while executing a work item handler it is not allowed to unready :)

bjarki-andreasen-nordic committed:
doc: kernel: serivces: threads: warn unsafe sys work queue use

Add warning to workqueue docs, explaining that using the system work
queue for blocking work can not be done safely.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
kernel: sys_workq: add k_is_in_sys_work API

Add k_is_in_sys_work() API which returns true if the thread context
is the system workqueue thread and the thread is currently servicing
a work item. Useful for checking if blocking is safe or required in
the case of calling an API which uses the system work queue
internally.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
kernel: sched: impl sys workq item no block enforcement

The system workqueue should never be unreadied from a work item
handler (while busy). This commit implements an optional check
which will invoke a kernel oops is a blocking operation is
attempted from a work item passed to the system workqueue.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

faxe1008 commented:
Very good catch, have found myself tracking down issues because of this multiple times :^)

bjarki-andreasen-nordic committed:
tests: kernel: workq: work_queue: use k_busy_wait to sim work

System workqueue items must not use blocking APIs, like k_msleep().
Replace k_msleep() with k_busy_wait() to both adhere to this rule,
and to better emulate more realistic work being done by system
workqueue.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

JordanYates commented:
For a more limited version of this change (Only preventing net buf blocking) that caused a chain of consequences and was ultimately reverted (in #84282), see #71697 (and the issues/PR's linked at the bottom).

IMO step 2 should be a separate PR. That looks like a nice change that can be applied independently of anything else.

cfriedt commented:
The system workqueue has generally been known as not the right workqueue for anything latency sensitive for a long time. I think it's been described as a kind of dumpster in the past.

Definitely agree that it shouldn't be used for anything latency sensitive, especially when those latencies can be infinite :-)

I'm ready to +1 this assuming we can get tests passing.

Knowingly using the system workqueue for anything blocking should really be an opt-in thing for very minimal designs / niche use-cases.

pdgendt commented on include/zephyr/kernel.h:
```diff
@@ -1181,6 +1181,19 @@ void k_thread_time_slice_set(struct k_thread *th, int32_t slice_ticks,
+ *
+ * @funcprops \isr_ok
+ *
+ * @return false if noit invoked from a system workqueue item.
```
```suggestion
 * @return false if not invoked from a system workqueue item.
```

bjarki-andreasen replied:
done

Thalley reviewed (commented):
Since this is relying on ASSERT, you may also want to consider that asserts are disabled by default, and that there may be many other places (including samples and tests) that are using the system workqueue in a blocking manor, which won't be caught be CI.

Besides the ASSERT you may also want to add a LOG_WRN or something which would shed more light on this issue

rugeGerritsen commented on kernel/Kconfig:
```diff
@@ -600,6 +600,15 @@ config SYSTEM_WORKQUEUE_NO_YIELD
 	  without yielding.
 
+config SYSTEM_WORKQUEUE_BLOCKING
+	bool "Select whether system work queue allows blocking calls"
```
The `default n` of SYSTEM_WORKQUEUE_BLOCKING is a change in behavior and may break existing applications. These applications may or may not have a risk of running into deadlocks.

Therefore, I believe we should consider setting the default to `y` and issue a warning for a deprecation period.





. Should there be a deprecation period where there is initially

bjarki-andreasen replied:
We can add `default y`, and deprecate it immediately, I'm all for that :)

Thalley replied:
> We can add `default y`, and deprecate it immediately, I'm all for that :)

Wouldn't that go against policy? i.e. no use of deprecated symbols in the Zephyr tree, so if set to `y` by default and it's deprecated, then it needs to be set to `=n` in all tests, samples, etc.?

bjarki-andreasen replied:
Yeah, I immediately noticed this would require disabling it everywhere manually :) not optimal

bjarki-andreasen commented on kernel/sched.c:
```diff
@@ -523,6 +523,9 @@ static inline void z_vrfy_k_thread_resume(k_tid_t thread)
 
 static void unready_thread(struct k_thread *thread)
 {
+#ifndef CONFIG_SYSTEM_WORKQUEUE_BLOCKING
+	__ASSERT_NO_MSG(!k_is_in_sys_work());
+#endif /* CONFIG_SYSTEM_WORKQUEUE_BLOCKING */
```
It gets a bit long and hard to read I think, but I can change it
```
__ASSERT_NO_MSG(!(!IS_ENABLED(CONFIG_SYSTEM_WORKQUEUE_BLOCKING) && k_is_in_sys_work()));
```
or do you mean
```
        if (!IS_ENABLED(CONFIG_SYSTEM_WORKQUEUE_BLOCKING)) {
                __ASSERT_NO_MSG(!k_is_in_sys_work());
        }
```
I have no strong opinion about this :)

Thalley replied:
I did mean
```
        if (!IS_ENABLED(CONFIG_SYSTEM_WORKQUEUE_BLOCKING)) {
                __ASSERT_NO_MSG(!k_is_in_sys_work());
        }
``` 

:)

bjarki-andreasen replied:
Code now looks like:
```
	if (IS_ENABLED(CONFIG_SYSTEM_WORKQUEUE_NO_BLOCK) && k_is_in_sys_work()) {
		k_oops();
	}
```
given additional feedback regarding use of k_oops instead of ASSERT :)

bjarki-andreasen commented on tests/kernel/mem_protect/demand_paging/mem_map/prj.conf:
```diff
@@ -5,3 +5,5 @@ CONFIG_DEMAND_PAGING_THREAD_STATS=y
 CONFIG_DEMAND_PAGING_TIMING_HISTOGRAM=y
 CONFIG_TEST_USERSPACE=y
 CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=0
+CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=0
```
ah, too fast on the ctrl+c ctrl-v :) I will fix it once CI has had a chance to run the tests :)

nashif commented on doc/kernel/services/threads/workqueue.rst:
```diff
@@ -103,6 +103,17 @@ operations that are potentially blocking (e.g. taking a semaphore) must be
+   potentially blocking, as there is no guarantee that work items submitted to
+   it do not depend on subsequent work items in the queue to unblock them.
+
+   Attempting to perform a blocking call will fail an assert if
```
we should not depend on ASSERTs and add config options to work  around them. Asserts should not be enable by default, only during development. This change and warning might give the impression that asserts should be enabled which is not correct.

Thalley replied:
> I see, I will remove the note, it may just be confusing. The config however is needed, since some rather complex subsystems like bluetooth do block the system workqueue at this time, which will cause an assert (if enabled). I can make the assert optional maybe, like `STACK_CANARIES` is used now, so you get the check if you need it, maybe enable it by default if `CONFIG_DEBUG=y`?

Would it be possible/worth to fix the places where Zephyr (incorrectly) blocks the system workqueue thread first, instead of trying to apply a workaround? 

I got the feeling last time this was discussed in the BT WG that we may want to make all BT APIs non-blocking. @jhedberg @alwa-nordic may have comments here. 

The change proposed by this PR is a significant change in behavior, and while it's OK to lets users apply fixes during a deprecation period, we should just fix the offending parts of Zephyr right away IMO. The challenge here is of course that if e.g. the BT subsystem changes behavior as well, then this deprecation period needs to be applied to all APIs that may be offending today.

bjarki-andreasen replied:
I do think upstream, like downstream users, should get the same grace period to align the use of the system work queue, looking at the last commit in this PR, there are quite a few places to fix, this PR may end up being quite large and unwieldy juggling changes to all of them :)

Thalley replied:
> this PR may end up being quite large and unwieldy juggling changes to all of them :)

Agreed. Depending on your expected timeframe, we could request the maintainers of the subsystems that needs to be modified, to modify those in the (very) near future in separate PRs. This PR can then act as a verifier of those. 

That being said, this PR could also go in first with it defaulting to the same behavior as now (i.e. allow blocking) and then the people fixing the places that may block can enable the option for debugging purposes, and then when all is green, we can change the default behavior (i.e. to not allow blocking).

bjarki-andreasen commented:
> I would like this proposal to take a step back and document what the desired end state is. The desire from users to defer tasks to some arbitrary point in the future still exists, and since blocking APIs are also things that exist, some of those tasks will want to use those APIs. What is the intended solution?

The end state is to not have the system workqueue be misused by any part if the kernel, and help users use it appropriately from their apps.

> Does everyone define their own "single user" workqueue? That seems sub-optimal.

Applications and subsystems can create workqueues to share internally, like RTIO, Bluetooth and soon PM. They are still shared but the scope is controlled.

> Will Zephyr define a second workqueue for tasks that need to block? Then the same reasoning applied to the system workqueue in this PR will apply to that queue as well.

No, it is fundamentally unsafe to do so.

> Unless the desired end state is the removal of all blocking APIs from Zephyr, I don't see the problem actually being solved by this PR. It will instead lead to either a RAM blowout, or the same problem occurring on a different work queue.

The issue this PR helps solve is in the description. No, the desired state is not to remove all blocking APIs. It will not lead to "RAM blowout". Other workqueus, like the one(s) used by RTIO, do call blocking APIs safely, it can be done in limited scopes.

JordanYates commented:
> The end state is to not have the system workqueue be misused by any part if the kernel, and help users use it appropriately from their apps.

I doubt the kernel is using the system work queue for blocking operations anywhere.
You are "helping" by mandating that they can't do something.

> Applications and subsystems can create workqueues to share internally, like RTIO, Bluetooth and soon PM. They are still shared but the scope is controlled.

Several of those subsystems (Bluetooth and RTIO) run user provided callbacks, is it left up to the subsystem to define whether user callbacks are allowed to block or not?

>> Unless the desired end state is the removal of all blocking APIs from Zephyr, I don't see the problem actually being solved by this PR.

> The issue this PR helps solve is in the description.

I should have been more explicit. I understand that this PR solves the surface level issue (blocking on the system work queue), but doesn't actually solve any  deadlocks itself. The exact same deadlocks can still exist, just on some other work queue. And the price to pay for this "kicking the problem somewhere else" is that the system work queue is generally less useful for everyone.

If every blocking user of the system work queue would be required to define their own work queue under this new regime, why not just create a dedicated work queue only for those subsystems known to be problematic, and leave everyone else to use the system work queue as it is?

JordanYates commented:
> I truly fail to understand why you are so opposed to this.

Because a much more limited version of this was merged for Bluetooth and it triggered months of issues and PRs before finally being reverted. I just don't want to see the same thing repeated here. I am perfectly okay with this happening if the consequences have been acknowledged.

IMO this should only be merged AFTER existing users have been transitioned to individual work queues, doing it before is just asking for problems.

bjarki-andreasen commented on kernel/sched.c:
```diff
@@ -523,6 +523,9 @@ static inline void z_vrfy_k_thread_resume(k_tid_t thread)
 static void unready_thread(struct k_thread *thread)
 {
+#ifndef CONFIG_SYSTEM_WORKQUEUE_BLOCKING
+	__ASSERT_NO_MSG(!k_is_in_sys_work());
```
Agreed :)

alwa-nordic commented:
> > Will Zephyr define a second workqueue for tasks that need to block? Then the same reasoning applied to the system workqueue in this PR will apply to that queue as well.
> 
> No, it is fundamentally unsafe to do so.

I agree. We should consider that [Zephyr has dynamic threads](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/kernel/threads/dynamic_thread/src/main.c). That looks like cheap threads that are perfect for short-lived tasks that need their own stack but don't want the static allocation.

pdgendt commented:
Question: If some high level call is done in the system work queue, that would cause a mutex/semaphore/... wait in some lower level (for example in an SPI driver), will this result in a kernel oops if the kconfig symbol isn't set?
I typically see no issue with these "blocks" as they should be freed after a very short delay.

IMO this will be confusing for new users. Also it's perfectly valid to wait in other threads, why should we make the system work queue special?

pdgendt commented on kernel/Kconfig:
```diff
@@ -600,6 +600,14 @@ config SYSTEM_WORKQUEUE_NO_YIELD
 	  without yielding.
 
+config SYSTEM_WORKQUEUE_NO_BLOCK
+	bool "Select whether system work queue enforces work items dont block"
+	help
+	  By default, the system work queue does not enforce work items
+	  passed to it dont perform blocking operations. Selecting this
+	  enforces that blocking operations are not performed by invoking
+	  a kernel oops if such operations are attempted.
```
```suggestion
	bool "Select whether system work queue enforces non-blocking work items"
	help
	  By default, the system work queue does not enforce work items
	  passed to it to not perform blocking operations. Selecting this
	  enforces that blocking operations are not performed by invoking
	  a kernel oops if such operations are attempted.
```

bjarki-andreasen replied:
Applied, will push once CI gets to run :)

bjarki-andreasen commented:
> Question: If some high level call is done in the system work queue, that would cause a mutex/semaphore/... wait in some lower level (for example in an SPI driver), will this result in a kernel oops if the kconfig symbol isn't set? I typically see no issue with these "blocks" as they should be freed after a very short delay.

If the symbol is not selected, no kernel oops :)

The issue with "blocks" here is that they have to be un-blocked by another context. If that context is another thread, and the block is very short, no problem. However, if the un-block needs to be performed by a later work item passed to the same work queue, there is a deadlock.

> IMO this will be confusing for new users. Also it's perfectly valid to wait in other threads, why should we make the system work queue special?

The system workqueue is already special, in that it has an "unlimited"/uncontrolled scope, compared to RTIO for example where every item passed to it is known and accounted for. Any module, driver, subsystem etc. on either side on an API call, can delegate work to the system workqueue, so there is no guarantee that the system workqueue is not both the calling context of a blocking call, and the context supposed to unblock it.

alwa-nordic commented:
> Question: If some high level call is done in the system work queue, that would cause a mutex/semaphore/... wait in some lower level (for example in an SPI driver), will this result in a kernel oops if the kconfig symbol isn't set? I typically see no issue with these "blocks" as they should be freed after a very short delay.
> 
> IMO this will be confusing for new users. Also it's perfectly valid to wait in other threads, why should we make the system work queue special?


I agree. This is a major problem.

Consider an operation that is extremely fast, but requires a large amount of memory. It would make sense to implement it using static memory protected by a mutex. This operation would never block for any significant amount of time, and would never participate in a deadlock. I think it's reasonable to treat the mutex as an implementation detail. It's completely safe to invoke this operation from the system work queue.

Then consider a library that previously implemented the above operation by allocating on the stack, but transitions to doing it the way described above. I think it would be reasonable for a library author to change it without considering it to be a breaking change. But, with this PR, it would start failing on the system work queue. This is very unfortunate for stability guarantees.

I think the terms 'blocking' and 'non-blocking' don't describe this situation well. This PR actually enforces that all work items on the system work queue are isr-ok, a stronger requirement, right?

bjarki-andreasen commented:
> > Question: If some high level call is done in the system work queue, that would cause a mutex/semaphore/... wait in some lower level (for example in an SPI driver), will this result in a kernel oops if the kconfig symbol isn't set? I typically see no issue with these "blocks" as they should be freed after a very short delay.
> > IMO this will be confusing for new users. Also it's perfectly valid to wait in other threads, why should we make the system work queue special?
> 
> I agree. This is a major problem. I want to expand on your point:
> 
> Consider an operation that is extremely fast, but requires a large amount of memory. It would make sense to implement it using static memory protected by a mutex. This operation would never block for any significant amount of time, and would never participate in a deadlock. I think it's reasonable to treat the mutex as an implementation detail. It's completely safe to invoke this operation from the system work queue.

In this case, use a spinlock. Accessing shared data is allowed from the system workqueue :) Mutexes and semaphores only become relevant once you start calling other APIs which you don't have control over.

> Then consider a library that previously implemented the above operation by allocating on the stack, but transitions to doing it the way described above. I think it would be reasonable for a library author to change it without considering it to be a breaking change. But, with this PR, it would start failing on the system work queue. This is very unfortunate for stability guarantees.

Should have been using a spinlock :)

> I think the terms 'blocking' and 'non-blocking' don't describe this situation well. This PR actually enforces that all work items on the system work queue are ISR-safe, a stronger requirement, right?

That is one way to put it : ) They need to be unblocking, which is also a requirement of ISR-safe code. The difference is you get a large stack at the cost of higher latency using the system work queue compared to actually calling something from ISRs :)

nashif commented on doc/kernel/services/threads/workqueue.rst:
```diff
@@ -103,6 +103,15 @@ operations that are potentially blocking (e.g. taking a semaphore) must be
 used with care, since the workqueue cannot process subsequent work items in
 its queue until the handler function finishes executing.
 
+.. warning::
```
why is this warning re system workqueue is being added here when we have a section dedicated to the system workqueue that already touches on the subject?

![image](https://github.com/user-attachments/assets/2973e5ef-81d1-426b-b798-6b60d8d0007c)

bjarki-andreasen replied:
Did not see this warning, it lacks the warning regarding deadlocking which is the crucial one :) I can move it to this section if we decide on continuing with this PR :)

teburd replied:
Isn't the deadlocking issue true of any work queue though? There's nothing particularly special about the system work queue other than it tends to get used by default a lot?

I think its worth pointing out a simple scenario where this occurs as well. 

E.g. one work item is taking a semaphore a subsequent work item is giving. Work queue is now dead locked.

bjarki-andreasen replied:
> Isn't the deadlocking issue true of any work queue though? There's nothing particularly special about the system work queue other than it tends to get used by default a lot?
> 
> I also think its worth pointing out a simple scenario where this occurs as well. 
> 
> E.g. one work item is taking a semaphore a subsequent work item is giving. Work queue is now dead locked.
> 
> Blocking calls aren't inherently the issue here either I'd note, its a possible symptom but not the *cause* of the deadlock.
> 
> A call to i2c_transfer() for example in a work queue item is a blocking call, and may cause the work queue thread to pend. Just because it blocks doesn't inherently mean there will be a deadlock!

The "which is available to any application or kernel code" part it what makes it true especially for the sys workq, given an owner of the queue would know all work passed to the queue, so can prevent deadlocks and manage latencies :)

peter-mitsis reviewed (approved):
This seems reasonable to me--particularly since it provides a means to allow a project to choose whether or not to allow blocking operations in the system work queue.

andyross commented on kernel/Kconfig:
```diff
@@ -600,6 +600,14 @@ config SYSTEM_WORKQUEUE_NO_YIELD
 
+config SYSTEM_WORKQUEUE_NO_BLOCK
+	bool "Select whether system work queue enforces non-blocking work items"
+	help
```
Maybe `default y if ASSERT` or something similar?  This is a cheap check with clear value, probably wants to be on any time CONFIG_ASSERT=y

andyross commented on kernel/sched.c:
```diff
@@ -523,6 +523,10 @@ static inline void z_vrfy_k_thread_resume(k_tid_t thread)
 static void unready_thread(struct k_thread *thread)
 {
+	if (IS_ENABLED(CONFIG_SYSTEM_WORKQUEUE_NO_BLOCK) && k_is_in_sys_work()) {
+		k_oops();
```
This looks wrong to me.  "Ready" and "running" aren't the same thing.  A thread can be ready but lower priority than _current.  Basically: my guess is that this code will oops if you try to `k_thread_suspend()` a runnable thread out of a work queue item, which would be expected to be legal and work.

You need to add a test for thread == _current at least, but it would probably be better to move this test to reschedule() instead.

Also: probably want a panic here and not an oops.  An oops in userspace will kill only the current thread, but a misuse of the system workqueue (which obviously is a kernel thread anyway) is a global failure.

And finally: neither oops nor panic give any feedback to the poor user whose code blew up.  Probably wants a printk() here (or to be expressed as an __ASSERT() when available).

andyross reviewed (changes requested):
Pretty sure the test is in the wrong spot.  Agree with the feature on the whole though.

Personally I'm neutral on whether this should apply to all work queues or just the system work queue.  Doesn't seem like there's much value to an app trying to block in its own queue, but at the same time it has clear and definable semantics and I guess there's no reason to disallow it.

jhedberg commented:
I think it's worth noting that one category of system workqueue "misuse" that this would not prevent, is when a work item doesn't block on any kernel object but still ends up blocking the workqueue itself due to spending lots of time doing "something", e.g. crypto operations lasting for multiple seconds, like we've seen in #84216.

JordanYates reviewed (changes requested):
> This seems reasonable to me--particularly since it provides a means to allow a project to choose whether or not to allow blocking operations in the system work queue.

I don't believe this leads to a choice at all. If there is an option to disable all blocking on the system workqueue, all in-tree system workqueue users need to avoid blocking. Otherwise you can't turn on the feature without the application breaking immediately.

IMO the only reliable way the choice can work is if it defaults to non-blocking in-tree, and downstream users can enable blocking only if their application needs it for whatever reason.

The problem is that you cannot turn this on by default, because there are too many in-tree users that rely on the blocking behavior. If I am wrong and you can enable this by default, I'm happy to give a +1. My 2 cents.

andyross commented:
> If there is an option to disable all blocking on the system workqueue, all in-tree system workqueue users need to avoid blocking. Otherwise you can't turn on the feature without the application breaking immediately.

Are there any such users?  I don't think it's unreasonable at all that we enforce "No Blocking The System Work Queue" rule for in-tree code even if we are more flexible for apps, in which case this PR would still have value.

brandon-exact commented:
if I understand this correctly, wouldnt this require a tree wide effort to change anything that blocks? For example, these GPIO drivers that would need to be updated

https://github.com/zephyrproject-rtos/zephyr/blob/1446416b6d9f6d86197acd25755a6abeb7bf5837/drivers/gpio/gpio_tca6424a.c#L180-L229
https://github.com/zephyrproject-rtos/zephyr/blob/1446416b6d9f6d86197acd25755a6abeb7bf5837/drivers/gpio/gpio_sx1509b.c#L189-L227
https://github.com/zephyrproject-rtos/zephyr/blob/1446416b6d9f6d86197acd25755a6abeb7bf5837/drivers/gpio/gpio_pcf857x.c#L85-L118
https://github.com/zephyrproject-rtos/zephyr/blob/1446416b6d9f6d86197acd25755a6abeb7bf5837/drivers/gpio/gpio_pcal64xxa.c#L198-L275
https://github.com/zephyrproject-rtos/zephyr/blob/1446416b6d9f6d86197acd25755a6abeb7bf5837/drivers/gpio/gpio_mcp23xxx.c#L405-L465

teburd commented:
> > If there is an option to disable all blocking on the system workqueue, all in-tree system workqueue users need to avoid blocking. Otherwise you can't turn on the feature without the application breaking immediately.
> 
> Are there any such users? I don't think it's unreasonable at all that we enforce "No Blocking The System Work Queue" rule for in-tree code even if we are more flexible for apps, in which case this PR would still have value.

Any fetch+get sensor with trigger handling optionally uses the "GLOBAL_THREAD" which is the system workqueue. Probably 50+ drivers.

bjarki-andreasen commented:
I have created an alternative to this PR which monitors for timeouts rather than if the work blocks, see #88345
